ElastAlert Repo Community Rule Support

coreyogburn opened the pull request:
Refactored Strelka and ElastAlert to share a GetReposDefault function for reading repos out of the config.

ElastAlert's new repos config allows for git repo support. Similar to Strelka, these community rules will be tagged with the repo they came from.

coreyogburn committed:
ElastAlert Repo Community Rule Support

Refactored Strelka and ElastAlert to share a GetReposDefault function for reading repos out of the config.

ElastAlert's new repos config allows for git repo support. Similar to Strelka, these community rules will be tagged with the repo they came from.

jertel commented on server/modules/elastalert/elastalert.go:
```diff
@@ -355,12 +382,20 @@ func (e *ElastAlertEngine) startCommunityRuleImport() {
+
+		repoDets, errMap := e.parseRepoRules(allRepos)
 		if errMap != nil {
 			log.WithField("error", errMap).Error("something went wrong while parsing sigma rule files")
```
Avoid using identical error strings, to better facilitate troubleshooting problems reported by customers.

coreyogburn replied:
Updated to be more specific.

jertel commented on server/modules/elastalert/elastalert.go:
```diff
@@ -88,6 +94,17 @@ func (e *ElastAlertEngine) Init(config module.ModuleConfig) error {
+	e.reposFolder = module.GetStringDefault(config, "reposFolder", "/opt/sensoroni/sigma/repos")
+	e.rulesRepos, err = module.GetReposDefault(config, "rulesRepos", []*module.RuleRepo{
+		{
+			Repo:    "https://github.com/Security-Onion-Solutions/securityonion-resources",
```
By making the securityonion-resources the default/fallback repo, it might not occur to people that choose to specify their own repo in the config that by doing so they are effectively skipping over rules from the SO repo. Would it make sense to move this default repo into the defaults.yaml, so that when they go to this property in configurator, they see it and can choose to replace it or add another to it?

jertel replied:
Nvm, I see now in the second PR you already list this as a default, so this is a fallback to someone removing the property completely. I think anyway. If they removed the SO repo from the configuration option would it still default to the SO repo here? If so this may be undesired, in the scenario where they explicitly want to stop pulling the SO rules. I think it comes down to how the GetReposDefault will behave when given a property containing an empty list.

coreyogburn replied:
There's a separate PR with an updated defaults.yaml: https://github.com/Security-Onion-Solutions/securityonion/pull/12640

coreyogburn replied:
It will only fallback in the case the config doesn't have the `rulesRepos` field present at all. If it's present, even if it's empty, that list from the config is used without modification.

coreyogburn committed:
Optional Folder in Detections Repos

Now you can specify a folder in ElastAlert and Strelka community rule repos. Without a folder, the entire repo is recursively scanned for rules. With a folder, the recursive scanning begins in the specified folder.

coreyogburn committed:
Unique Error Strings
